File: 2/p1.go
package main

import "io/ioutil"
import "fmt"
import "strconv"
import "strings"

func handlePosition (p []int, pos int) ([]int, int) {
    op_code := p[pos]

    if op_code == 99 {
        return p, -1
    }

    a_pos := p[pos + 1]
    a := p[a_pos]
    b_pos := p[pos + 2]
    b := p[b_pos]
    c_pos := p[pos + 3]

    switch op_code {
        case 1:
            p[c_pos] = a + b
        case 2:
            p[c_pos] = a * b
    }

    return p, pos + 4
}

func parseFile (fileName string) []int {
    data, _ := ioutil.ReadFile(fileName)
    program := []int{}
    for _, s := range strings.Split(string(data), ",") {
        i, err := strconv.Atoi(strings.Trim(s, "\n"))
        if err == nil {
            program = append(program, i)
        }
    }
    return program
}

func run (program []int) []int {
    position := 0
    for position >= 0 {
        program, position = handlePosition(program, position)
    }
    return program
}

func main () {
    program := parseFile("program.intcode")
    program[1] = 12
    program[2] = 2
    fmt.Println(run(program)[0])
}

File: 2/p1.py
def handle_position(program, position):
    op_code = program[position]
    if op_code == 99:
        return program, -1

    a_pos = program[position + 1]
    a = program[a_pos]

    b_pos = program[position + 2]
    b = program[b_pos]

    c_pos = program[position + 3]

    if op_code == 1:
        program[c_pos] = a + b
    if op_code == 2:
        program[c_pos] = a * b

    return program, position + 4

def run_program(program):
    position = 0
    while position >= 0:
        program, position = handle_position(program, position)

    return program


def parse_file(filename):
    with open(filename) as f:
        code = f.readline()
        ints = code.split(',')
        program = [int(i) for i in ints]
        return program


program = parse_file('program.intcode')
program[1] = 12
program[2] = 2
result = run_program(program)

print result[0]
